constructors.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package config
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+)
+
+type scopedProvider struct {
+	Provider
+
+	prefix string
+}
+
+var _ Provider = (*scopedProvider)(nil)
+
+func (s *scopedProvider) Get(key string) Value {
+	return s.Provider.Get(fmt.Sprintf("%s.%s", s.prefix, key))
+}
+
+// NewScopedProvider wraps a provider and adds a prefix to all Get calls.
+func NewScopedProvider(prefix string, provider Provider) Provider {
+	if prefix == "" {
+		return provider
```

Comment:
Should we create a new provider anyways? The signature implies that we are creating a new instance so this would remove any possibility of mutation (e.g. a custom `config.Provider` implementation from the user).

```
func NewScopedProvider(prefix string, provider Provider) Provider {
    return &scopedProvider{provider, prefix /* can be "" */}
}
```

Reply:
I don't think that's necessary; if the provider implementation mutates its internal state when users call the exported API, we'll have problems anyways (since the `scopedProvider` is just delegating to the user-supplied provider).

Reply:
👍